File: app/models/shipment.py
import uuid
from datetime import UTC, datetime

from sqlalchemy import DateTime, Float, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Shipment(Base):
    """A shipment from one node to another (depot → site)."""

    __tablename__ = "shipments"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    from_node_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("inventory_nodes.id"), index=True)
    to_node_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("inventory_nodes.id"), index=True)

    status: Mapped[str] = mapped_column(String(32), default="REQUESTED", index=True)
    # REQUESTED → APPROVED → PICKED → SHIPPED → IN_TRANSIT → RECEIVED → CANCELLED

    lane_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    tracking_number: Mapped[str | None] = mapped_column(String(256), nullable=True)
    temperature_req: Mapped[str | None] = mapped_column(String(64), nullable=True)  # e.g. "2-8C", "AMBIENT"
    courier: Mapped[str | None] = mapped_column(String(128), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    requested_by: Mapped[str | None] = mapped_column(String(128), nullable=True)
    requested_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    approved_by: Mapped[str | None] = mapped_column(String(128), nullable=True)
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    shipped_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    received_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    items: Mapped[list["ShipmentItem"]] = relationship(back_populates="shipment", cascade="all, delete-orphan")


class ShipmentItem(Base):
    """A line item in a shipment — specific lot + qty."""

    __tablename__ = "shipment_items"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    shipment_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("shipments.id"), index=True)
    lot_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("inventory_lots.id"))

    product_id: Mapped[str] = mapped_column(String(128))
    presentation_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    qty: Mapped[float] = mapped_column(Float)

    shipment: Mapped["Shipment"] = relationship(back_populates="items")

File: frontend/src/types/shipment.ts
import type { UUID } from './common';

export interface ShipmentItem {
  id: UUID;
  shipment_id: UUID;
  lot_id: UUID;
  product_id: string;
  presentation_id: string | null;
  qty: number;
}

export interface Shipment {
  id: UUID;
  from_node_id: UUID;
  to_node_id: UUID;
  status: string;
  lane_id: string | null;
  tracking_number: string | null;
  temperature_req: string | null;
  courier: string | null;
  notes: string | null;
  requested_by: string | null;
  requested_at: string;
  approved_by: string | null;
  approved_at: string | null;
  shipped_at: string | null;
  received_at: string | null;
  items: ShipmentItem[];
}

export interface ShipmentCreate {
  from_node_id: string;
  to_node_id: string;
  items: { lot_id: UUID; product_id: string; presentation_id?: string; qty: number }[];
  lane_id?: string;
  temperature_req?: string;
  courier?: string;
  notes?: string;
  requested_by?: string;
}

export interface ShipmentAction {
  performed_by?: string;
  tracking_number?: string;
  notes?: string;
}

export type ShipmentStatus = 'REQUESTED' | 'APPROVED' | 'PICKED' | 'SHIPPED' | 'IN_TRANSIT' | 'RECEIVED' | 'CANCELLED';
